Add support for μ-law encoded PCM to scipy.io.wavfile.   This implementation decodes to 16-bit PCM.

ksmathers opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://docs.scipy.org/doc/numpy/dev/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
N/A

#### What does this implement/fix?
Adds support for MULAW PCM formatted audio files to scipy.io.wavfile

#### Additional information
<!--Any additional information you think is important.-->

tylerjereddy commented on scipy/io/wavfile.py:
```diff
@@ -304,7 +305,7 @@ class WAVE_FORMAT(IntEnum):
 
 
-KNOWN_WAVE_FORMATS = {WAVE_FORMAT.PCM, WAVE_FORMAT.IEEE_FLOAT}
+KNOWN_WAVE_FORMATS = {WAVE_FORMAT.PCM, WAVE_FORMAT.IEEE_FLOAT, WAVE_FORMAT.MULAW }
```
The addition of the new format means that `test_read_unknown_wave_format()` will fail because mulaw is now handled. Not sure if there's another known algorithm that could be substituted in to that unit test until we support it, or if we should just get rid of that test if all major formats are supported now. Could also just add a completely invalid format type I suppose.

ksmathers replied:
There are a long list of WAV formats that aren't supported by scipy, but MULAW and ALAW are interesting because they are both trivial enough to be implemented on an 8-bit microcontroller and compressed enough to transfer over a UART or almost any other serial device like an XBee or similar wireless IoT networks in realtime.

tylerjereddy commented:
Oh yeah, out of curiosity, I noticed mention of a "copyright" issue in one of the commits?

larsoner commented:
@endolith can you comment?

ksmathers commented:
My original commit used code from another project on GitHub.  Looking
through the overall project I couldn't find any evidence that the code was
licensed at all much less for compatibility with the Berkeley license, so I
replaced it with a simple look up table which should be faster anyway.  The
code for looking up values in that table is my own, as are the various
other added bits, but the reference wasn't to any complaint it was just
that I didn't want to wait for confirmation from the other author that I
could reuse his code, so I replaced it.

On Tue, Aug 18, 2020 at 6:59 PM Eric Larson <notifications@github.com>
wrote:

> @endolith <https://github.com/endolith> can you comment?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scipy/scipy/pull/12735#issuecomment-675807141>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAC3SXBWXYYK4ZOOF4LJZY3SBMWWLANCNFSM4QDXRZVA>
> .
>

endolith commented:
I don't know much about μ-law.  What bit depths are supported?  The bit depth checking will be changed to handle different formats separately in https://github.com/scipy/scipy/pull/12287/files#diff-c5c9583b55bf7580040d99e32c4f1771R429  Does this mesh well with that?

ksmathers commented:
Compliant MULAW recordings are restricted to 8-bit.   The conversion back
to PCM in my code masks the values to 8-bits if they exceed that so you
could give it data in larger bit depths and it would continue to treat that
data as G.711 MULAW 8-bit representation but really the meaning is
undefined, it just doesn't throw exceptions.

On Tue, Aug 18, 2020 at 8:48 PM endolith <notifications@github.com> wrote:

> I don't know much about μ-law. What bit depths are supported? The bit
> depth checking will be changed to handle different formats separately in
> https://github.com/scipy/scipy/pull/12287/files#diff-c5c9583b55bf7580040d99e32c4f1771R429
> Does this mesh well with that?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scipy/scipy/pull/12735#issuecomment-675835179>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAC3SXGK3CWVLQAKKFNVV6LSBNDQLANCNFSM4QDXRZVA>
> .
>

endolith commented:
Now that #12287 is merged, can you rebase onto that?  The line `if format_tag == WAVE_FORMAT.PCM or format_tag == WAVE_FORMAT.MULAW:` will need to be split out into its own check, so 

* if PCM:
  * check PCM bit depths that we support
* if float
  * check float bit depths that we support
* if mulaw
  * check 8 bit that we support
